npm/pkg/dataplane/dpshim/dpshim.go

```diff
@@ -1,77 +1,590 @@
-func (dp *DPShim) InitializeDataPlane() error {
-	return nil
+	return &DPShim{
+		outChannel:  outChannel,
+		setCache:    make(map[string]*controlplane.ControllerIPSets),
+		policyCache: make(map[string]*policies.NPMNetworkPolicy),
+		dirtyCache:  newDirtyCache(),
+	}, nil
 }
 
 func (dp *DPShim) ResetDataPlane() error {
 	return nil
 }
 
-func (dp *DPShim) RunPeriodicTasks() {}
+// HydrateClients is used in DPShim to hydrate a restarted Daemon Client
+func (dp *DPShim) HydrateClients() (*protos.Events, error) {
+	dp.Lock()
+	defer dp.Unlock()
+
+	if len(dp.setCache) == 0 && len(dp.policyCache) == 0 {
+		klog.Infof("HydrateClients: No local cache objects to hydrate daemon client")
+		return nil, nil
+	}
+
+	goalStates := make(map[string]*protos.GoalState)
+
+	toApplySets, err := dp.hydrateSetCache()
+	if err != nil {
+		return nil, err
+	}
+	if toApplySets != nil {
+		goalStates[controlplane.IpsetApply] = toApplySets
+	}
+
+	toApplyPolicies, err := dp.hydratePolicyCache()
+	if err != nil {
+		return nil, err
+	}
+	if toApplyPolicies != nil {
+		goalStates[controlplane.PolicyApply] = toApplyPolicies
+	}
+
+	if len(goalStates) == 0 {
+		klog.Info("HydrateClients: No changes to apply")
+		return nil, nil
+	}
+
+	return &protos.Events{
+		Payload: goalStates,
+	}, nil
+}
+
+func (dp *DPShim) RunPeriodicTasks() {
+	// Here Run periodic task to check if any sets with empty references are present and delete them
+}
 
 func (dp *DPShim) GetIPSet(setName string) *ipsets.IPSet {
 	return nil
 }
 
+func (dp *DPShim) getIPSet(setName string) *controlplane.ControllerIPSets {
+	return dp.setCache[setName]
+}
+
+func (dp *DPShim) setExists(setName string) bool {
+	_, ok := dp.setCache[setName]
+	return ok
+}
+
 func (dp *DPShim) CreateIPSets(setNames []*ipsets.IPSetMetadata) {
+	dp.Lock()
+	defer dp.Unlock()
+	for _, set := range setNames {
+		dp.createIPSet(set)
+	}
+}
+
+func (dp *DPShim) createIPSet(set *ipsets.IPSetMetadata) {
+	setName := set.GetPrefixName()
+
+	if dp.setExists(setName) {
+		return
+	}
+
+	dp.setCache[setName] = controlplane.NewControllerIPSets(set)
+	dp.dirtyCache.modifyAddorUpdateSets(setName)
 }
 
 func (dp *DPShim) DeleteIPSet(setMetadata *ipsets.IPSetMetadata) {
+	dp.Lock()
+	defer dp.Unlock()
+	dp.deleteIPSet(setMetadata)
 }
 
-func (dp *DPShim) AddToSets(setNames []*ipsets.IPSetMetadata, podMetadata *dataplane.PodMetadata) error {
-	return nil
+func (dp *DPShim) deleteIPSet(setMetadata *ipsets.IPSetMetadata) {
+	setName := setMetadata.GetPrefixName()
+	klog.Infof("deleteIPSet: cleaning up %s", setName)
+	set, ok := dp.setCache[setName]
+	if !ok {
+		return
+	}
+
+	if set.HasReferences() {
+		return
+	}
+
+	delete(dp.setCache, setName)
+	dp.dirtyCache.modifyDeleteSets(setName)
 }
 
-func (dp *DPShim) RemoveFromSets(setNames []*ipsets.IPSetMetadata, podMetadata *dataplane.PodMetadata) error {
+func (dp *DPShim) AddToSets(setMetadatas []*ipsets.IPSetMetadata, podMetadata *dataplane.PodMetadata) error {
+	dp.Lock()
+	defer dp.Unlock()
```

Comment:
If we have this in `createIPSet` we wouldn't have to lock on a large code block

Reply:
except for the exists check everything needs to be locked anyway, so i believe trying to lock for each loop might add more overhead. wdyt ?